### PR TITLE
[DTensor] Strategy Validation (3/3): strategy querying, orchestrator, and CLI

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #174995
* __->__ #174800
* #174799

Adds the orchestrator (compare_operator) that ties everything together:
queries DTensor for its claimed sharding rules via three strategy paths
(single-dim, op_strategy, decomp), computes ground truth validity for
each placement combination, and reports discrepancies (incorrect rules
and missing rules).

Includes false positive mitigations (sign negation for P(min)/P(max),
non-rounded variants for rounding_mode ops), a CLI entry point for
running validation on individual ops or all registered ops, and
end-to-end tests.

### Example Usage:

`python -m torch.distributed.tensor._ops.strategy_validation --op add,mul --max 1 --show-repro`
```
Testing ops: aten.add, aten.mul
Device: cuda, Dtype: torch.float32, World size: 2

[1/2] aten.add — Samples: 1, Combinations: 120
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.add.Tensor]
    P(avg), R -> P(avg)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')
    P(max), R -> P(max)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')
    P(min), R -> P(min)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')
    R, P(avg) -> P(avg)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')
    R, P(max) -> P(max)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')
    R, P(min) -> P(min)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')

[2/2] aten.mul — Samples: 1, Combinations: 120
----------------------------------------------------------------------

Possibly missing (valid in ground truth but no DTensor rule)

  [aten.mul.Tensor]
    R, P(avg) -> P(avg)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')
    R, P(sum) -> P(sum)
      Repro: self=tensor(-6.7103, device='cuda:0'), other=tensor(2.1750, device='cuda:0')

======================================================================
Summary
======================================================================
Op        Correct  Incorrect  Missing    Time
---------------------------------------------
aten.add        2          0        6     1.9s
aten.mul        2          0        2     1.6s
---------------------------------------------
Total           4          0        8     3.5s
```

### Basic design: 
Screenshot 2026-02-02 at 2 38 56 PM

**DTensor Incorrect** should be reliably detected: any report of incorrect by this tool should be a DTensor bug.

**DTensor Missing** rules is inherently less reliable:
- these are detected by finding cases where a particular placement gives correct outputs, and this can be data-dependent (can trigger false positives) - e.g. if the sample input was all 0, we could expect any partial placement to work. 
- This PR already includes significant work towards de-noising partials- it creates local values that are not equal to each other but reduce to the correct global value. This weeds out many false positives in my limited testing.
- This de-noising infra can continue to be enhanced as new cases are encountered.

### CLI:
` python -m torch.distributed.tensor._ops.strategy_validation -h`
<details><summary>

```
usage: strategy_validation.py [-h] [--op OP] [--all-registered] [--incorrect-only] [--device DEVICE] [--dtype DTYPE] [--world-size WORLD_SIZE] [--max-samples MAX_SAMPLES]
                              [--show-repro [N]]
```

</summary>

```
Compare DTensor rules against ground truth

options:
  -h, --help            show this help message and exit
  --op OP               Operator name(s) to compare (comma-separated, supports glob patterns, e.g., "relu,add" or "nn.functional.*")
  --all-registered      Test all ops with DTensor sharding rules registered
  --incorrect-only      Only test DTensor's claimed rules (faster, skips missing detection)
  --device DEVICE       Device to use
  --dtype DTYPE         Dtype to use
  --world-size WORLD_SIZE
                        Simulated world size
  --max-samples MAX_SAMPLES
                        Max samples to test
  --show-repro [N]      Show N sample repros per rule (default 1 if flag given, -1 for all)
```
</details>

Authored with Claude.

### CODE REVIEW

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/tensor/_ops/strategy_validation.py [hunk 1]
```diff
@@ -524,3 +567,889 @@ def negate_tensor(x):
...
+            expanded_result.append(expanded_combo)
+
+        return expanded_result
+    except Exception:
+        return None
+
+
+def get_opinfo_by_name(name: str) -> list[opinfo_core.OpInfo]:
+    """Find OpInfo entries by operator name."""
+    matches = [op for op in op_db if op.name == name]
```
Comment (wconstab) [inline torch/distributed/tensor/_ops/strategy_validation.py:727]: @anshul-si suggested that aten.relu doesn't exist in opinfo but functional relu does, so probably this function needs to be improved

File: torch/distributed/tensor/_ops/strategy_validation.py [hunk 2]
```diff
@@ -548,3 +591,889 @@ def negate_tensor(x):
...
+                continue
+
+            if any(0 in t.shape for _, t in tensors):
+                continue
+
+            total_samples += 1
+
+            try:
+                ground_truth = _run_op_on_sample(op, sample)
+                if not isinstance(ground_truth, torch.Tensor):
```
Comment (zpcore) [inline torch/distributed/tensor/_ops/strategy_validation.py:1274]: Do we care the List[Tensor] output like split()?

File: torch/distributed/tensor/_ops/strategy_validation.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/strategy_validation.py:727]: i have improved this: the script now has more handling for finding ops.  relu now works, as does dropout (which used to not find samples, but now finds them and is explicitly skipped since its a random op cc @zpcore)

File: torch/distributed/tensor/_ops/strategy_validation.py (hunk 2 above)
Comment (wconstab) [reply torch/distributed/tensor/_ops/strategy_validation.py:1274]: i have first made the script clearer about 'skip reasons' so we know why it got skipped.  i will then experiment with supporting the non-tensor output ops and see if it's clean or maybe do it in a separate PR

**Review by zpcore — APPROVED** [review]
I can't think of those corner cases from opinfo. The logic looks good!
